File: Var10CSharp/Var10Cshar/WaterBottles.cs
/*
class WaterBottles
{
    static void Main()
    {
        // Ввод количества банок с водой
        Console.Write("Введите количество банок с водой: ");
        int bottles = int.Parse(Console.ReadLine());
        // Ввод курса обмена пустых банок на новые
        Console.Write("Введите курс обмена пустых банок на новые: ");
        int rate = int.Parse(Console.ReadLine());

        // Вызов функции DrinkWater для определения максимального числа выпитых банок
        int bottlesDrank = DrinkWater(bottles, rate);
        // Вывод результата
        Console.WriteLine("Максимальное число банок, которое можно выпить: " + bottlesDrank);
    }

    // Функция DrinkWater для вычисления максимального числа выпитых банок
    static int DrinkWater(int bottles, int rate)
    {
        int bottlesDrank = bottles; // Изначально количество выпитых банок равно начальному количеству
        int emptyBottles = bottles; // Изначально количество пустых банок равно начальному количеству
        int steps = 1; // Изначально количество шагов равно 1

        // Вывод информации о начальном состоянии
        Console.WriteLine("Дано: ");
        PrintHappyFaces(bottles);
        Console.WriteLine("Шаг: " + steps);
        PrintEmptyBottles(emptyBottles);

        // Пока количество пустых банок больше или равно курсу обмена
        while (emptyBottles >= rate)
        {
            int exchangedBottles = emptyBottles / rate; // Вычисление количества обмененных банок
            bottlesDrank += exchangedBottles; // Увеличение общего числа выпитых банок
            int remainingEmptyBottles = emptyBottles % rate; // Оставшиеся пустые банки после обмена
            emptyBottles = exchangedBottles + remainingEmptyBottles; // Обновление количества пустых банок
            steps++; // Увеличение количества шагов

            // Вывод информации о текущем состоянии
            Console.WriteLine("Шаг: " + steps);
            PrintHappyFaces(exchangedBottles);
            PrintEmptyBottles(remainingEmptyBottles);

            steps++; // Увеличение количества шагов
            Console.WriteLine("Шаг: " + steps);
            PrintEmptyBottles(emptyBottles);
        }

        // Вывод информации о количестве шагов и возврат общего числа выпитых банок
        Console.WriteLine("Количество итераций: " + steps);
        return bottlesDrank;
    }

    // Функция для вывода символов "☺", обозначающих выпитые банки
    static void PrintHappyFaces(int n)
    {
        for (int i = 1; i <= n; i++)
        {
            Console.Write("☺");
            if (i % 3 == 0)
            {
                Console.WriteLine();
            }
            if (i == n && i % 3 != 0)
            {
                Console.WriteLine();
            }
        }
    }

    // Функция для вывода символов "©", обозначающих пустые банки
    static void PrintEmptyBottles(int n)
    {
        for (int i = 1; i <= n; i++)
        {
            Console.Write("©");
            if (i % 3 == 0)
            {
                Console.WriteLine();
            }
            if (i == n && i % 3 != 0)
            {
                Console.WriteLine();
            }
        }
    }
}
*/

File: Var10Ruby/water_bottles.rb
def main
  puts "Введите количество банок с водой: "
  bottles = gets.chomp.to_i  # Ввод количества банок с водой

  puts "Введите курс обмена пустых банок на новые: "
  rate = gets.chomp.to_i  # Ввод курса обмена пустых банок на новые

  bottles_drank = drink_water(bottles, rate)  # Вызов функции для определения максимального числа банок, которое можно выпить
  puts "Максимальное число банок, которое можно выпить: #{bottles_drank}"  # Вывод результата
end

def drink_water(bottles, rate)
  bottles_drank = bottles  # Инициализация переменной с количеством выпитых банок
  empty_bottles = bottles  # Инициализация переменной с количеством пустых банок
  steps = 1  # Инициализация переменной с количеством шагов

  puts "Дано: "  # Вывод заголовка
  print_happy_faces(bottles)  # Вызов функции для вывода смайликов
  puts "Шаг: #{steps}"  # Вывод номера текущего шага
  print_empty_bottles(empty_bottles)  # Вызов функции для вывода пустых банок

  while empty_bottles >= rate do  # Проверка, есть ли еще пустые банки для обмена
    exchanged_bottles = empty_bottles / rate  # Вычисление количества банок для обмена
    bottles_drank += exchanged_bottles  # Увеличение количества выпитых банок
    remaining_empty_bottles = empty_bottles % rate  # Вычисление оставшихся пустых банок
    empty_bottles = exchanged_bottles + remaining_empty_bottles  # Обновление количества пустых банок
    steps += 1  # Увеличение числа шагов

    puts "Шаг: #{steps}"  # Вывод номера текущего шага
    print_happy_faces(exchanged_bottles)  # Вызов функции для вывода смайликов
    print_empty_bottles(remaining_empty_bottles)  # Вызов функции для вывода пустых банок

    steps += 1  # Увеличение числа шагов
    puts "Шаг: #{steps}"  # Вывод номера текущего шага
    print_empty_bottles(empty_bottles)  # Вызов функции для вывода пустых банок
  end

  puts "Количество итераций: #{steps}"  # Вывод количества итераций
  bottles_drank  # Возвращаем количество выпитых банок
end

def print_happy_faces(n)
  print "☺" * n  # Вывод смайликов
  puts
end

def print_empty_bottles(n)
  print "©" * n  # Вывод пустых банок
  puts
end

main  # Вызов основной функции
